ARCHITECTURE.md

```diff
@@ -0,0 +1,221 @@
+
+Finally, PythonClass can also generate a separate internal representation using `namedtuple`s to allow browsing CBOR data by the names given in the CDDL.
+(This is more analogous to how the data is accessed in the C code.)
+
+PythonClass functionality is tested in [tests/scripts](tests/scripts)
+
+CodeGenerator
+-------------
+
+CodeGenerator, like PythonClass, inherits from CddlXcoder.
+Its primary purpose is to construct the individual decoding/encoding functions for the types specified in the given CDDL document.
+It also constructs struct definitions used to hold the decoded data/data to be encoded.
+
+CodeGenerator contains optimizations to reduce both the verbosity of the code and the level of indirection in the types.
+For example:
+ - If the type is unambiguous (i.e. its value is predetermined, like in `Foo = 5`), the code will validate it, but CodeGenerator won't include the actual value in the encompassing struct definition.
+ - If a `"GROUP"` or `"UNION"` has only one child, it can be removed as a level of indirection.
+ - If the type needs only a single member variable (i.e. no additional `foo_count` or `foo_key` etc.), that variable can instead be added to the parent struct, and its decoding/encoding code moved into the parent's function.
+ - `"UNION"` are typically implemented as anonymous `union`s which removes one level of indirection when accessing them.
+
+A CodeGenerator object operates in one of two modes: `"encode"` or `"decode"`.
+The generated code for the two is not very different, but they call into different libraries.
+
+Base types, like `"UINT"`, `"BOOL"`, `"FLOAT"` are represented by native C types. `"BSTR"`s and `"TSTR"`s are represented by a proprietary `cbor_string_type_t` which is just a `uint8_t` pointer and length.
+These types are decoded/encoded with C code that is not generated.
+More on this in the Architecture of the generated C code below.
+
+When a type is repeated (maxQ > 1 or maxQ > minQ), there needs to be a distinction between repeated_foo() and foo() (these can be either encoding or decoding functions).
+repeated_foo() concerns itself with the individual value, while foo() concerns itself with the value including repetitions.
+
+When invoking CodeGenerator, the user must decide which types it will need direct access to decode/encode.
+These are typically the "outermost" types, or the types it is expected that the data will have, and they are kept in `self.entry_types`.
+
+The user can also use Entry types when there are `"BSTR"`s that are CBOR encoded, specified as `Foo = bstr .cbor Bar`.
+Usually such strings are automatically decoded/encoded by the generated code, and the objects part of the encompassing struct.
+However, sometimes it's desirable that these are not automatically decoded/encoded, so if they are found in `self.entry_types` they will be stored as a regular `cbor_string_type_t`, and the user must manually decode/encode it via the provided API.
```

Comment:
```suggestion
However, sometimes it is desirable that these are not automatically decoded/encoded; storing them as regular `cbor_string_type_t` types if present in `self.entry_types` and manually decoding/encoding them via the provided API.
```
?

Reply:
```suggestion
However, if the user instead wants to manually decode/encode such strings, they can add them to `self.entry_types`, in which case they will be stored as a regular `cbor_string_type_t` instead, and not decoded/encoded.
```

Reply:
I like it! Some minor tweaks:

```suggestion
However, if the user instead wants to manually decode/encode such strings, they can add them to `self.entry_types`. In this case, the strings will be stored as a regular `cbor_string_type_t` instead of being decoded/encoded.
```